pkg/controller/status.go

```diff
@@ -32,6 +32,20 @@ func (c *Controller) ClusterStatus(team, cluster string) (*spec.ClusterStatus, e
 	return status, nil
 }
 
+// GetClusterDatabasesMap returns for each cluster the list of databases running there
+func (c *Controller) GetClusterDatabasesMap() map[string][]string {
+
+	m := make(map[string][]string)
+
+	for _, cluster := range c.clusters {
```

Comment:
This is dangerous: c.clusters is not a syncMap, and concurrent writes to the map itself may happen (i.e. if a new cluster is add). Concurrent read/writes to the map may either corrupt the map or produce a panic.

There is a cluster.Mu mutex that you should acquire (by calling RLock) before accessing c.clusters. Note that you should not hold this mutex for too long, as it is used in many other places, some of them are driven by events coming from Kubernetes. Therefore, the right way would be to:
- call c.cluster.Mu.Rlock()
- read all keys from the map
- call c.cluster.Mu.RUnlock()
- for each key from the list you've just read
  - call c.cluster.Mu.Rlock()
  - assign the c.cluster[$key] to the local variable
  -  call c.cluster.Mu.RUnlock()

after that you can access the cluster.Postgresql.Spec.Databases from your local variable.

Reply:
Actually, it's a bit more complex: even if we copy the cluster to the local map, the Spec.Databases, which is a map, will only reference the original one (and, potentially, be prone to the same issues with concurrent map read/writes). What we can do instead of assigning the cluster to the local variable is iterating through the c.cluster[$key].Postgresql.Spec.Databases, copying them to the local map (while being protected by the c.cluster.Mu mutex).

Reply:
Is there a reason we dont just use a concurrent map? Probably does not mage an impact on performance.

Reply:
The only reason is that we started before the concurrent map came in place, so we might consider it here, although just a concurrent map won't solve our issues, since we also access the cluster-data that can be concurrently changed at another place (i.e. when the cluster is updated) and are guarded by a separate lock (which is held for the whole time the cluster operation happens, so at the end of the day the API call may wait a lot). One of the ideas is to introduce a separate mutex lock for the spec changes, so that the API code will only interfere with only a few functions when the cluster state is changed.

Reply:
The database map is it really modified or replaced by a new map? In which case would be fine. I cannot find the code right now.

Reply:
Both the database map is modified (It is, for instance, when we add new clusters or delete old ones), and the individual map elements (i.e. updating the cluster or syncing it) could be changed as well. That's why we need to mutexes, one for the map itself and other for every cluster.